File: plugins/todo.1m.py
#!/usr/bin/env python3
"""Todo List - Simple todo manager"""
import json
import os

TODO_FILE = os.path.expanduser('~/.crossbar/todos.json')

def load_todos():
    try:
        if os.path.exists(TODO_FILE):
            with open(TODO_FILE, 'r') as f:
                return json.load(f)
    except:
        pass
    return []

def save_todos(todos):
    os.makedirs(os.path.dirname(TODO_FILE), exist_ok=True)
    with open(TODO_FILE, 'w') as f:
        json.dump(todos, f, indent=2)

todos = load_todos()
pending = [t for t in todos if not t.get('done', False)]
done = [t for t in todos if t.get('done', False)]

count = len(pending)
icon = '' if count == 0 else ''

print(f"{icon} {count}")
print("---")

if pending:
    print("Pending:")
    for i, todo in enumerate(pending):
        text = todo.get('text', 'Untitled')
        print(f"  {text}")
else:
    print("No pending tasks")

if done:
    print("---")
    print("Completed:")
    for todo in done[:5]:  # Show last 5
        text = todo.get('text', 'Untitled')
        print(f"  {text}")

print("---")
print(f"Total: {len(todos)} | Pending: {len(pending)} | Done: {len(done)}")
print("---")
print("Refresh | refresh=true")

File: plugins/todo/todo.1m.lua
-- todo.1m.lua
-- Simple Todo List using text file

local function env(name, default)
    local value = crossbar.env(name, default)
    if value == nil or value == '' then
        return default
    end
    return value
end

local function expand_home(path)
    if path:sub(1, 1) == '~' then
        return crossbar.homeDir() .. path:sub(2)
    end
    return path
end

local function shell_quote(value)
    local escaped = tostring(value):gsub("'", "'\\''")
    return "'" .. escaped .. "'"
end

local todo_file = expand_home(env('TODO_FILE', crossbar.homeDir() .. '/.crossbar/todo.txt'))

local function read_todos()
    local f = io.open(todo_file, "r")
    if not f then return {} end
    local lines = {}
    for line in f:lines() do
        table.insert(lines, line)
    end
    f:close()
    return lines
end

local todos = read_todos()
local count = #todos

local icon = count > 0 and 'TODO' or 'DONE'

print(icon .. ' ' .. count)
print("---")

if count > 0 then
    for _, task in ipairs(todos) do
        print(task)
    end
else
    print('No pending tasks')
end

print("---")
local open_cmd = "crossbar open file " .. shell_quote(todo_file)
print('Edit Todo List | bash="' .. open_cmd .. '" terminal=false refresh=true')
print("Refresh | refresh=true")
